draft-ietf-tls-external-psk-importer.md

```diff
@@ -79,156 +80,151 @@ when, and only when, they appear in all capitals, as shown here.
+If the EPSK is a key derived from some other protocol or sequence of protocols,
+ImportedIdentity.context MUST include a channel binding for the deriving protocols
+{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included
+in ImportedIdentity.context {{CCB}}.
 
-ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
-from some other protocol or sequence of protocols, ImportedIdentity.context MUST include a channel binding for the deriving protocols
-{{!RFC5056}}.  If any secrets are agreed in earlier protocols they SHOULD be included in ImportedIdentity.context [CCB].
+ImportedIdentity.target_protocol MUST be the (D)TLS protocol version for which the
+PSK is being imported. For example, TLS 1.3 {{!RFC8446}} and QUICv1 {{!I-D.ietf-quic-transport}}
+MUST use 0x0304. Note that this means future versions of TLS will increase the number of PSKs
+derived from an external PSK.
 
-A unique and imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:
+An imported PSK (IPSK) with base key 'ipskx' bound to this identity is then computed as follows:
 
 ~~~
    epskx = HKDF-Extract(0, epsk)
    ipskx = HKDF-Expand-Label(epskx, "derived psk",
-                             Hash(ImportedIdentity), Hash.length)
+                             Hash(ImportedIdentity), L)
 ~~~
 
-[[TODO: The length of ipskx MUST match that of the corresponding and supported ciphersuites.]]
+L is the length of the hash function associated with ImportedIdentity.target_kdf, as this
+is required for the imported PSK to be of suitable length with compatible ciphersuites.
+The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK.
+It is not the hash function associated with ImportedIdentity.target_kdf. If no hash function
+is specified, SHA-256 MUST be used. Differentiating epsk by ImportedIdentity.target_kdf ensures
+that an imported PSK is only used as input keying material to at most one KDF, thus satisfying
+the requirements in {{!RFC8446}}.
 
-The hash function used for HKDF {{!RFC5869}} is that which is associated with the external PSK. It is not
-bound to ImportedIdentity.hash. If no hash function is specified, SHA-256 MUST be used.
-Differentiating epsk by ImportedIdentity.hash ensures that each imported PSK is only used with at most one
-hash function, thus satisfying the requirements in {{!RFC8446}}. Endpoints MUST import and derive an ipsk
-for each hash function used by each ciphersuite they support. For example, importing a key for
-TLS_AES_128_GCM_SHA256 and TLS_AES_256_GCM_SHA384 would yield two PSKs, one for SHA256 and another
-for SHA384. In contrast, if TLS_AES_128_GCM_SHA256 and TLS_CHACHA20_POLY1305_SHA256 are supported,
-only one derived key is necessary.
+Endpoints MUST import an ipskx for each target KDF they support. For example, importing a key for
```

Comment:
```suggestion
Endpoints generate an Imported PSK for each target KDF they support. For example, importing a key for
```

I don't think that you want a MUST here.  If I don't *want* to use this EPSK with a given KDF, I shouldn't be forced to.  For instance, if I have an EPSK that is "strong", I might not want to use it with a weaker KDF and fall back to ECDH instead.  That sounds crazy now that I write it down, but I don't think we can proscribe that behaviour.

Reply:
This should be for each ciphersuite. I'll drop the normative MUST and swap "for each target KDF" with "for each offered ciphersuite". (One shouldn't use a PSK that won't work with a ciphersuite offered on the wire.)